flatpak-import.py

```diff
@@ -0,0 +1,174 @@
+    for key in metadata_keys:
+        try:
+            metadata[key] = metadata_variant[key]
+        except KeyError:
+            metadata[key] = ''
+        logging.debug("{0}: {1}".format(key, metadata[key]))
+
+    # Open repository
+    repo_file = Gio.File.new_for_path(args.repo)
+    repo = OSTree.Repo(path=repo_file)
+    if os.path.exists(os.path.join(args.repo, 'config')):
+        logging.info('Opening repo at ' + args.repo)
+        repo.open()
+    else:
+        logging.info('Creating archive-z2 repo at ' + args.repo)
+        os.makedirs(args.repo)
+        repo.create(OSTree.RepoMode.ARCHIVE_Z2)
+
+    try:
+        # Prepare transaction
+        repo.prepare_transaction(None)
+        repo.transaction_set_ref(None, metadata['ref'], commit)
+
+        # Execute the delta
+        flatpak_file = Gio.File.new_for_path(args.flatpak)
+        repo.static_delta_execute_offline(flatpak_file, False, None)
+
+        # Verify gpg signature
+        if args.gpg_homedir:
+            gpg_homedir = Gio.File.new_for_path(args.gpg_homedir)
+        else:
+            gpg_homedir = None
+        if args.keyring:
+            trusted_keyring = Gio.File.new_for_path(args.keyring)
+        else:
+            trusted_keyring = None
+        gpg_verify_result = repo.verify_commit_ext(commit,
+                                                   keyringdir=gpg_homedir,
+                                                   extra_keyring=trusted_keyring,
+                                                   cancellable=None)
+        # TODO: Handle signature requirements
+        if gpg_verify_result and gpg_verify_result.count_valid() > 0:
+            logging.info("valid signature found")
+        else:
+            raise Exception("no valid signature")
+
+        # Commit the transaction
+        repo.commit_transaction(None)
+
+    except:
+        repo.abort_transaction(None)
+        raise
+
+    # Grab commit root
+    ret, commit_root, commit_checksum = repo.read_commit(metadata['ref'], None)
+    if not ret:
+        logging.critical("commit failed")
+        exit(1)
+    else:
+        logging.debug("commit_checksum: " + commit_checksum)
+
+    # Compare installed and header metadata, remove commit if mismatch
+    metadata_file = Gio.File.resolve_relative_path (commit_root, "metadata");
+    # TODO: GLib-GIO-CRITICAL **: g_file_input_stream_query_info: assertion
+    # 'G_IS_FILE_INPUT_STREAM (stream)' failed
+    ret, metadata_contents, _ = metadata_file.load_contents(cancellable=None)
+    if ret:
+        if metadata_contents == metadata['metadata']:
+            logging.debug("committed metadata matches the static delta header")
+        else:
+            logging.critical("committed metadata does not match the static delta header")
+            repo.set_ref_immediate(remote=None,
+                                   ref=metadata['ref'],
+                                   checksum=None,
+                                   cancellable=None)
```

Comment:
I'll try it. For some reason flatpak does the metadata comparison outside of the transaction.